🇬🇧 [English](02_modeling.ipynb) · 🇪🇸 [Español](02_modeling_es.ipynb)

# PBC cohort: modeling analysis

This notebook uses the maintained production modules in `src/pbc/` for loading,
splitting, preprocessing, fitting, threshold selection, and metrics. It is
runnable from either `cirrhosis/` or `cirrhosis/notebooks/`.

The primary endpoint is **Stage 3–4 versus Stage 1–2**, framed as *non-invasive staging to
avoid a biopsy*: every model is compared with the published Mayo risk score and APRI, on two
deployable feature sets (`core`: 412 patients, routine variables; `full`: the 312 trial patients
with the extended lab panel), using a rule-out threshold (sensitivity ≥ 90%). Exact four-stage
and cumulative ordinal models are secondary. This is exploratory research with no
external or temporal validation: **not for clinical use, not medical advice,
and not a replacement for biopsy or clinician assessment**. See
[AASLD PBC guidance](https://www.aasld.org/practice-guidelines/primary-biliary-cholangitis)
and [EASL PBC guidance](https://easl.eu/publication/management-of-cholestatic-liver-diseases/).

In [1]:
import json
from pathlib import Path

import numpy as np
import pandas as pd

_here = Path.cwd().resolve()
PROJECT_ROOT = next(p for p in [_here, *_here.parents]
                    if (p / "src" / "pbc" / "data.py").exists() and (p / "data" / "raw" / "pbc.csv").exists())

from pbc.data import FEATURE_SETS, events_per_variable, load_pbc_data, train_test_split_pipeline
from pbc.evaluation import (
    binary_metrics,
    bootstrap_metric_intervals,
    calibration_metrics,
    multiclass_metrics,
    select_operating_threshold,
)
from pbc.modeling import build_named_pipeline, fit_model
from pbc.scores import SCORES

In [2]:
DATA_PATH = PROJECT_ROOT / "data" / "raw" / "pbc.csv"
frame = load_pbc_data(DATA_PATH)
ML_MODELS = ("logistic", "random_forest", "hist_gradient_boosting")
def models_for(split):
    """ML comparators plus every clinical score whose inputs the feature set contains."""
    return [*ML_MODELS, *[n for n, (_, required) in SCORES.items() if set(required) <= set(split.X_train.columns)]]
print(f"Project root: {PROJECT_ROOT}; validated rows={len(frame)}")
print(f"ML models: {ML_MODELS}; clinical score baselines: {tuple(SCORES)}")

Project root: /home/datakrdo/Documents/portfolio/cirrhosis; validated rows=418
ML models: ('logistic', 'random_forest', 'hist_gradient_boosting'); clinical score baselines: ('mayo', 'apri')


The notebook resolves its root from the presence of `src/pbc/data.py` and the
validated CSV, so execution location cannot silently select a different
dataset. The `pbc` package is imported from the project environment (`uv sync`).

Reproducibility matters because model claims from a small historical cohort
are especially sensitive to data provenance.

Only the production loaders and estimators are used; core preprocessing or
model logic is never copied into notebook cells.

In [3]:
splits = {}
for feature_set in FEATURE_SETS:
    split = train_test_split_pipeline(frame, feature_set=feature_set, test_size=0.2, random_state=41)
    splits[feature_set] = split
    print(f"[{feature_set}]", {
        "predictors": list(split.X_train.columns),
        "train_rows": len(split.X_train), "test_rows": len(split.X_test),
        "train_test_index_overlap": len(set(split.X_train.index) & set(split.X_test.index)),
        "events_per_variable_train": round(events_per_variable(split.X_train, split.y_binary_train), 3),
    })
    print("  Train endpoint counts:", split.y_binary_train.value_counts().sort_index().to_dict(),
          "| Test endpoint counts:", split.y_binary_test.value_counts().sort_index().to_dict())

[core] {'predictors': ['Age', 'Sex', 'Edema', 'Bilirubin', 'Albumin', 'Platelets', 'Prothrombin'], 'train_rows': 329, 'test_rows': 83, 'train_test_index_overlap': 0, 'events_per_variable_train': 12.857}
  Train endpoint counts: {0: 90, 1: 239} | Test endpoint counts: {0: 23, 1: 60}
[full] {'predictors': ['Age', 'Sex', 'Edema', 'Bilirubin', 'Albumin', 'Platelets', 'Prothrombin', 'Ascites', 'Hepatomegaly', 'Spiders', 'Cholesterol', 'Copper', 'Alk_Phos', 'SGOT', 'Tryglicerides'], 'train_rows': 249, 'test_rows': 63, 'train_test_index_overlap': 0, 'events_per_variable_train': 4.4}
  Train endpoint counts: {0: 66, 1: 183} | Test endpoint counts: {0: 17, 1: 46}


Each feature set is split once (80/20), stratified on the endpoint: `core` uses all 412 labeled
patients, `full` only the 312 trial patients, the only ones with the extended panel. `Drug` and
`trial_cohort` are never predictors (`Drug` proxies cohort membership almost perfectly; `trial_cohort`
does not exist for a new patient — see `01_eda.ipynb`), and follow-up time and status can encode
disease outcome after baseline, so they are excluded too. Events-per-variable on the training fold is
12.9 for `core` and 4.4 for `full`; the latter is below the conventional EPV >= 10 rule of thumb,
which is reported rather than hidden.

All imputers/encoders/models are fit on training rows only, the operating threshold is selected by
cross-validation on the training fold only (no manual validation split), and the test set is evaluated
once.

In [4]:
thresholds, final_models, test_probabilities = {}, {}, {}
for feature_set, split in splits.items():
    thresholds[feature_set], final_models[feature_set], test_probabilities[feature_set] = {}, {}, {}
    for name in models_for(split):
        pipeline = build_named_pipeline(name, split.X_train, random_state=41)
        threshold, tuned = select_operating_threshold(
            pipeline, split.X_train, split.y_binary_train.to_numpy(), random_state=41
        )
        thresholds[feature_set][name] = threshold
        final_models[feature_set][name] = tuned
        test_probabilities[feature_set][name] = tuned.predict_proba(split.X_test)[:, 1]
    print(f"[{feature_set}] CV-tuned rule-out thresholds (max specificity s.t. sensitivity >= 90%):")
    print({k: round(v, 4) for k, v in thresholds[feature_set].items()})

[core] CV-tuned rule-out thresholds (max specificity s.t. sensitivity >= 90%):
{'logistic': 0.5382, 'random_forest': 0.5493, 'hist_gradient_boosting': 0.487, 'mayo': 0.5992}


[full] CV-tuned rule-out thresholds (max specificity s.t. sensitivity >= 90%):
{'logistic': 0.5051, 'random_forest': 0.5424, 'hist_gradient_boosting': 0.4871, 'mayo': 0.5433, 'apri': 0.5578}


All ML models and both clinical scores are trained through `build_named_pipeline` (fresh
fold-fitted preprocessing per model — median-impute/one-hot for `logistic`/`random_forest`, no imputation
and native categorical dtype for `hist_gradient_boosting`; for `mayo`/`apri`, the fixed score formula followed by
a one-feature logistic recalibration). There is no `class_weight`, so predicted probabilities keep the cohort's
73% prevalence instead of being pushed toward 50%.

The operating threshold comes from `select_operating_threshold`, a rule-out objective: maximize specificity
subject to sensitivity >= 90%, chosen by cross-validation on the training fold only. No test label ever touches
threshold selection, and no model is selected based on test performance; the deployed model is `logistic`,
pre-registered for parsimony, calibration and explainability.

In [5]:
held_out = {}
for feature_set, split in splits.items():
    y_test = split.y_binary_test.to_numpy()
    held_out[feature_set] = {}
    for name, probabilities in test_probabilities[feature_set].items():
        metrics = binary_metrics(y_test, probabilities, threshold=thresholds[feature_set][name])
        metrics["bootstrap_95_ci"] = bootstrap_metric_intervals(
            y_test, probabilities, threshold=thresholds[feature_set][name], n_bootstrap=100, random_state=41
        )
        metrics["calibration"] = calibration_metrics(y_test, probabilities)
        held_out[feature_set][name] = metrics
    summary = pd.DataFrame(held_out[feature_set]).T[["auroc", "auprc", "sensitivity", "specificity", "npv", "biopsies_avoided", "brier_score"]]
    summary["calibration_slope"] = [m["calibration"]["calibration_slope"] for m in held_out[feature_set].values()]
    print(f"\n[{feature_set}] held-out test fold, n={len(y_test)}, prevalence={y_test.mean():.3f}:")
    print(summary.astype(float).round(3).to_string())
artifact = PROJECT_ROOT / "outputs" / "notebook_modeling_metrics.json"
artifact.parent.mkdir(exist_ok=True)
artifact.write_text(json.dumps({"primary_endpoint": "Stage 3-4 vs Stage 1-2", "feature_sets": held_out}, indent=2) + "\n")
print(f"\nWrote {artifact}")


[core] held-out test fold, n=83, prevalence=0.723:
                        auroc  auprc  sensitivity  specificity    npv  biopsies_avoided  brier_score  calibration_slope
logistic                0.618  0.822        0.950        0.043  0.250             0.048        0.198              0.614
random_forest           0.636  0.848        0.900        0.043  0.143             0.084        0.196              0.758
hist_gradient_boosting  0.689  0.865        0.983        0.000  0.000             0.012        0.189              1.037
mayo                    0.701  0.867        0.900        0.261  0.500             0.145        0.181              1.373



[full] held-out test fold, n=63, prevalence=0.730:
                        auroc  auprc  sensitivity  specificity    npv  biopsies_avoided  brier_score  calibration_slope
logistic                0.774  0.912        0.935        0.176  0.500             0.095        0.166              1.239
random_forest           0.789  0.911        0.891        0.235  0.444             0.143        0.163              1.499
hist_gradient_boosting  0.735  0.887        0.891        0.235  0.444             0.143        0.174              1.049
mayo                    0.753  0.901        0.935        0.176  0.500             0.095        0.168              1.512
apri                    0.664  0.836        1.000        0.000  0.000             0.000        0.189              1.814

Wrote /home/datakrdo/Documents/portfolio/cirrhosis/outputs/notebook_modeling_metrics.json


AUROC, AUPRC, sensitivity, specificity, NPV, the share of patients a rule-out would spare a biopsy
(`biopsies_avoided`) and percentile bootstrap intervals are computed on the untouched test predictions. On
the `core` test fold (n=83) the ML models' AUROCs range 0.62–0.69, and the Mayo score
(0.70) is at least as good as any of them; on `full` (n=63) they range
0.74–0.79 against 0.75 for Mayo and 0.66 for APRI. Intervals are
wide because only about one fifth of the labeled rows are held out, and the rule-out operating points on a single
fold are erratic (specificity from ~0 to ~0.3) — which is why `03_model_development.ipynb` is the source of
the headline numbers.

Uncertainty and class imbalance make a single point estimate unsafe for clinical decisions, especially in a
treatment-era cohort without external validation. None of this is used or deployed for patient care.

In [6]:
# Secondary endpoints use the core split (all 412 patients).
core = splits["core"]
secondary = {}
multiclass_model = fit_model("multiclass_logistic", core.X_train, core.y_stage_train, random_state=41)
secondary["multiclass_logistic"] = multiclass_metrics(
    core.y_stage_test.to_numpy(), np.asarray(multiclass_model.predict(core.X_test))
)
ordinal_model = fit_model("ordinal_logistic", core.X_train, core.y_stage_train, random_state=41)
secondary["ordinal_logistic"] = multiclass_metrics(
    core.y_stage_test.to_numpy(), np.asarray(ordinal_model.predict(core.X_test))
)
print(json.dumps(secondary, indent=2))
payload = json.loads(artifact.read_text())
payload["secondary"] = secondary
artifact.write_text(json.dumps(payload, indent=2) + "\n")

{
  "multiclass_logistic": {
    "macro_f1": 0.2610176282051282,
    "balanced_accuracy": 0.30833333333333335,
    "quadratic_weighted_kappa": 0.13947173537398172
  },
  "ordinal_logistic": {
    "macro_f1": 0.25939477848101267,
    "balanced_accuracy": 0.30833333333333335,
    "quadratic_weighted_kappa": 0.16098839385997754
  }
}


16008

Exact four-stage and cumulative ordinal estimators use the same core split and production
preprocessing, and are summarized with macro F1, balanced accuracy, and quadratic-weighted kappa rather than
the binary threshold.

Stage ordering carries clinical meaning, but adjacent-stage ambiguity and sparse early classes limit
certainty; ordinal agreement is not proof of valid staging.

These analyses stay secondary and hypothesis-generating; the binary endpoint remains the sole primary claim.

In [7]:
fitted = final_models["core"]["logistic"].estimator_
feature_names = list(fitted.named_steps["preprocess"].get_feature_names_out())
coefficients = fitted.named_steps["model"].coef_[0]
importance = pd.Series(coefficients, index=feature_names).sort_values(key=np.abs, ascending=False)
print("Logistic coefficients, core set (direction is model association, not causation):")
print(importance.round(3).to_string())
print("\nNote: hist_gradient_boosting exposes neither coef_ nor "
      "feature_importances_; its explanations come from SHAP in 03_model_development.ipynb.")
limitations = [
    "418 source rows, 412 labeled Stage values; the extended lab panel exists only for the 312 trial patients (structural, not random, missingness).",
    "Historical 1974-1984 Mayo cohort; spectrum and treatment-era transportability are unknown.",
    "Events-per-variable on the training fold is below the EPV >= 10 rule of thumb, especially for the full set.",
    "Single random holdout with no temporal or external validation; nested-CV estimates in 03_model_development.ipynb are the more defensible headline numbers.",
    "The Mayo score was derived on this same cohort (for survival), so it is a favourable, not external, comparator.",
    "No biopsy replacement, diagnosis, treatment recommendation, or clinical deployment.",
]
print("\nLimitations:")
for item in limitations:
    print(f"- {item}")

Logistic coefficients, core set (direction is model association, not causation):
Platelets     -0.357
Bilirubin      0.263
Edema          0.243
Prothrombin    0.236
Age            0.166
Albumin       -0.107
Sex_F          0.040
Sex_M         -0.039

Note: hist_gradient_boosting exposes neither coef_ nor feature_importances_; its explanations come from SHAP in 03_model_development.ipynb.

Limitations:
- 418 source rows, 412 labeled Stage values; the extended lab panel exists only for the 312 trial patients (structural, not random, missingness).
- Historical 1974-1984 Mayo cohort; spectrum and treatment-era transportability are unknown.
- Events-per-variable on the training fold is below the EPV >= 10 rule of thumb, especially for the full set.
- Single random holdout with no temporal or external validation; nested-CV estimates in 03_model_development.ipynb are the more defensible headline numbers.
- The Mayo score was derived on this same cohort (for survival), so it is a favourable, no

Coefficient rankings summarize one fitted model after one split and are
sensitive to correlated laboratory variables, missingness, and sample size;
they are not a stable feature-importance claim.

PBC staging requires clinical context and validated diagnostic pathways;
AASLD/EASL guidance, clinician judgment, and appropriate investigations
supersede this exploratory model.

These coefficients are used only to generate research hypotheses. External and
temporal validation are required before any translational
discussion, and this analysis is not a biopsy replacement or for clinical use.